src/callbacks/callback_learning_rate.cpp

```diff
@@ -212,6 +213,50 @@ float lbann_callback_linear_growth_learning_rate::global_schedule(model *m) {
   }
 }
 
+lbann_callback_poly_learning_rate::lbann_callback_poly_learning_rate(
+  double p, uint64_t n_epochs, uint64_t max_iter)
+  : lbann_callback_learning_rate(std::unordered_set<weights *>()),
+    m_p(p), m_num_epochs(n_epochs), m_max_iter(max_iter),
+    m_lr(1.0f), m_last_epoch_lr(1.0f) {}
+
+lbann_callback_poly_learning_rate::lbann_callback_poly_learning_rate(
+  double p, uint64_t n_epochs, uint64_t max_iter, std::unordered_set<weights *> weights_list)
+  : lbann_callback_learning_rate(weights_list),
+    m_p(p), m_num_epochs(n_epochs), m_max_iter(max_iter),
+    m_lr(1.0f), m_last_epoch_lr(1.0f) {}
+
+/**
+ * Check if the maximum number of iterations is set. If not, compute it by the
+ * number of epochs and the number of iterations per epoch.
+ */
+void lbann_callback_poly_learning_rate::setup(model *m) {
+  lbann_callback_learning_rate::setup(m);
+  if (m_max_iter == 0ull) {
+    m_max_iter = m_num_epochs * m->get_num_iterations_per_epoch(execution_mode::training);
+  }
+}
+
+/**
+ * Keep the record of the learning rate at the end of the current epoch.
+ */
+float lbann_callback_poly_learning_rate::global_schedule(model *m) {
+  const float amt = m_lr / m_last_epoch_lr;
+  m_last_epoch_lr = m_lr;
+  return m_cur_global_lr * amt;
+}
```

Comment:
This should be deleted.

Reply:
One thing that was not clear to me was whether optimizers are supposed to operate exclusively.

Reply:
It looks like they can all interact via the static variable `m_last_epoch_lr`. This interface seems very fragile. I'd prefer if each learning rate were independent, but @ndryden might be doing some fancy stuff with other learning rate schedules.